### DISCUSSION TITLE
User Badge system

### DISCUSSION THREAD

**Discussion by JaminShanti** [jamulussoftware/jamulus#2064] [Ideas]
I've noticed something using Jamulus over a month of usage.  It's hard to build rapport, like you would do on stackoverflow.  Maybe some sort of badge system would point out community members, talented community members, and new members.  Any interest in this?  Every time I login I feel like I need to re-introduce myself.  What's the long term vision for Jamulus?  I have found Open Source Software is all about popularity.  The popular ones get bigger, the others just fade away.

**Comment by chrisrimple** [comment]
Jamulus is actually an older project (10 years, I think) that gained new attention and life during the Covid pandemic.  It has mostly been the work of a single person (@corrados) but has added a number of contributors in the last 15 months, and is now maintained/advanced by a group of volunteers.  Would Jamulus benefit from having a "community"?  Possibly a community of developers, although we arguably already have that here.  There's also many communities of users - see the Facebook groups "Jamulus WorldJam", "Jamulus Choral Community", and others.

All that said, I'm unclear how badges or other community features would manifest in Jamulus.  There's no central list of users, no leaderboard, etc.  So I would only see your badge and understand your contribution level when we happen to be connected to the same server at the same time.  Given that most musicians connect to servers closest to them (lowest latency), they're generally seeing the same musicians over and over, a tiny subset of the total Jamulus users - is there really any benefit to badges?

I'm not dismissing the idea, just trying to understand the benefits.

**Reply by gilgongo** [reply]
https://lounge.jamulus.io (unfortunately down at the moment due to the theft of copper cabling at the ISP, apparently!) is an example of a "community" system for Jamulus where peoples' "metadata" might be displayed. It's also possible that some tweaks to Jamulus could be made to help interactions (eg [URL handling](https://github.com/jamulussoftware/jamulus/issues/494)) to make it easier to get to servers), but @chrisrimple makes some good points. 

Also when discussions of things like client avatars and notes comes up (not that these are being suggested here necessarily though) I'm told that adding extra data to the client's use of the UDP stream might degrade the audio too, but I'm not qualified to know.

**Comment by JaminShanti** [comment]
I'd like to know the number of long term users, I wonder if there is a group of constant users or is the usage design constant replacement.

**Reply by chrisrimple** [reply]
> I'd like to know the number of long term users, I wonder if there is a group of constant users or is the usage design constant replacement.

When you say "constant users" do you mean developers working on the open source code or musicians playing on Jamulus?  In the developer group, there's a small cadre of dedicated developers that have been contributing to Jamulus for the last year+, and continue to do so.  Occasionally a new developer shows up and wants to contribute, usually with a specific interest (Android client, iOS client, etc.).  In the musician group, there are a number of constant users - examples include the [Jamulus WorldJam](http://worldjam.vip), an almost-weekly streamed performance by musicians from around the world, "virtual bands" that have formed during the pandemic (search for I.P. Rockers on Facebook), and even local musicians that are using Jamulus as an alternative to rehearsing in person (my band rehearses twice a week via Jamulus).  It's hard to count the number of musicians utilizing Jamulus, but if we include the many choral groups that are having regular rehearsals, I would guess that it's in the thousands every week - most are using private servers so not visible to others.  While many of those groups will return to in-person rehearsals post-Covid, the virtual bands and some locals may continue to use Jamulus, and the WorldJam may continue, since it's provided a means for musicians from thousands of miles apart to perform together.

**Comment by JaminShanti** [comment]
User would be customers if it was a paid product. Introducing a badge system solves a problem, community retention. If you feel the community is the right size, no need for a badge system. "As far as I can see" is not the best metric. Bias to personal experience. Community members or users could be staying based on the quality of the solution, Jamulus is a good solution. On the other hand if you have continuous replacement, it might be less apparent that even though the number is at say 60, it's a different 60 every year. This means you have high attrition rates and a user bage system would help. I'm not trying to defeat my own post, but I suggested a solution for problem. The problem isn't perceived to exist no need for the solution.

**Reply by gene96817** [reply]
I  don't know if I like the idea of "customers". I do like the idea of building a community. There are a couple of tendencies that make building a community difficult.

1- Public vs Private servers.... some Jamulus users prefer (and recommend) using private servers over public servers. Private server communities are invisible. If enough communities go invisible, it will be harder to sustain the developer community.

2- Privacy -- some individuals have a very strict requirement for privacy... it would be great to have a system for aliases that meets the privacy requirement and still be useful for community building. It could be argued that consistent use of alias will assist in piercing privacy. Unfortunately, the rogue element that disrupts sessions relies on privacy to inflict their bad behavior on others.

That being said, community building is a worthy endevor.

**Reply by JaminShanti** [reply]
I would suggest the following.
As a group decide if community growth is on the wish list. If it is introduce a badge system. Badge systems don't attract new musicians, they retain existing musicians. The technical feasibility of a badge system sounds like it blocks many other design choices. So there decide how many improvements are being blocked by the limitations of audio connections over UDP are tracking musicians. When you have enough improvements that require this change implement the change.

**Reply by chrisrimple** [reply]
Fundamentally, Jamulus is a tool, just like a hammer - the "community" of users will never be larger than the total number of people needing/using a hammer.  WRT Jamulus, that community is further divided by music interest, skill, public/private, and possibly even location, not unlike hammer users: carpenters, woodworkers, hobbyists, etc.  Take a look at the FB group "Jamulus Choral Community" and you'll see lots of questions and answers from choral singers and directors wanting to use Jamulus and learning how to do so, some with tremendous success - for the most part, "community" has developed around Jamulus without the app (hammer) facilitating it.  Could Jamulus do more to facilitate "community"?  Certainly, and some ideas have been suggested in the past, including:

- Registry/directory of users, including location, skill level, music interest, etc.
- Leaderboard showing most active servers (last day, week, month), active times, etc.
- Registry/directory of "events" including group rehearsals, performances, etc. that are open to listeners
- Registry/directory of Jamulus users willing to help others get started, resolve tech issues, etc.

And so on.  While badges might be an interesting idea, I think Jamulus faces 3 large issues WRT to increasing the community size: awareness, getting started, and finding other musicians.  There's a fair number of good resources online for getting started, but there hasn't been worldwide media coverage to create awareness and there aren't good means to easily find other musicians with whom to play (within a reasonable distance/latency).  Solving those items would likely grow the community, even without adding community features to Jamulus itself.

**Reply by gilgongo** [reply]
The ideas of community functions and "Jamulus as a tool" are interesting because (a bit like Harley Davidson motorbikes or personalised Zippo lighters) there is a possibility that some form of secondary market could be built around it. In fact I assume this is the idea with that site that I always forget the name of - where you can spin up a private Jamulus server on demand and invite others to it. Still leaves the issues of getting started and setup though, which currently are big problems.

**Comment by JaminShanti** [comment]
As I write this post, it's 8pm EST.  I think there are 2 over 5 people groups online in the entire world publicly.

**Reply by chrisrimple** [reply]
You say that like it's a bad thing.

Seriously, I'm unclear what problem you're trying to solve?  That, as a musician, it's hard to find others with whom to jam/play?  That, as a new Jamulus musician, it may not be inviting to login regularly and find no one making music?  That, as a potential listener, it's hard to find music when you look?

As I said above, Jamulus is a hammer (tool) - it's not a Twitch stream.  It was never intended to be a listening tool, a performance tool, etc.  Yes, it's been used for those purposes successfully, but often in conjunction with other tools (Zoom, OBS, etc.).  Should the development team build out features in Jamulus to enable other use cases, or just refer interested musicians to instructions and tools to fill the gap?

**Reply by JaminShanti** [reply]
Sry how many commits to the project do you have?  I can't seem to find any.
https://github.com/jamulussoftware/jamulus/graphs/contributors

**Reply by chrisrimple** [reply]
@JaminShanti Correct, I have zero commits.  But I'm a tester for new releases, have contributed many ideas and responses (search my name in Issues and Bugs), have reviewed and provided feedback on UX designs, have answered countless questions and solved configuration issues for users, etc.  I'm also a member of the Admin Team, a very active volunteer, and have performed 50+ songs with musicians from around the world on the [Jamulus WorldJam](http://worldjam.vip), which has dramatically helped to promote Jamulus.  Perhaps most importantly, I'm the author and maintainer of the [Remote Band Rehearsals](https://docs.google.com/document/d/1smcvsxdaaViPQvGMQHmah_6BQeqowhmGSFMHfnlY2FI/) google doc, which has likely guided more musicians in getting started with Jamulus than any other source.  You might notice that it's referenced at the top of the official [Jamulus Tips & Tricks](https://jamulus.io/wiki/Tips-Tricks-More) page.

Since your interest seems to be in "community", perhaps you should spend a bit more time getting to know THIS ONE rather than talking out of your ass.

**Reply by gilgongo** [reply]
> As I write this post, it's 8pm EST. I think there are 2 over 5 people groups online in the entire world publicly.

Despite what else is implied by this statement, it is notable that large numbers of empty servers are a thing. It's hard to know why this is beyond the fact that of course for any one player the list of viable servers (measured by ping rate) will be different to another. My own guess is that when it comes to public jamming, few players want to be the first to enter a server, so the system relies on a very small number of "enterprising" players willing to hang out on their own until somebody else arrives. Addressing this is what lounge.jamulus.io is about - perhaps we should advertise this more (once it's back up)?

Either way, I think there is more "community" traction to be had outside of Jamulus than within the application itself.

**Reply by JaminShanti** [reply]
Imaging your senerio with a badge system.   Random server has 1 person online and second person joins.  They have a random profile instrument assigned, many times these images don't even match what comes from their mic. Does the batch system increase the chance a random 1 user server increasing or decrease it, or make no difference at all.

**Reply by ignotus666** [reply]
How do you know whether a person has randomly assigned themselves an instrument until you hear them? Would you dismiss someone up front before hearing them just because they don't have a particular badge? Ok, a certain badge can tell you that a person has a level of "trustworthiness" - but that automatically entails that a new user _doesn't_ have that credibility. Why? Based on what? On a forum, no one will stop you from posting as a new user regardless, unless you cross a line, enabling you to have a say without being ignored. But here it could lead to people being unfairly dismissed, just because they're "new" - when they could be magnificent musicians who are far more fun to play with than any "established" member. What I'm getting at is that a badge system says **nothing** about connection quality and musicianship, which I think is what most people value.

**Reply by npostavs** [reply]
>  Does the batch system increase the chance a random 1 user server increasing or decrease it, or make no difference at all.

I guess it would make no difference at all. I don't see how it could affect this scenario.

**Comment by ignotus666** [comment]
To me this "badge system" sounds a bit like a hierarchy, playing on people's desire to rise within it and creating a community more akin to a "member's club" that people can aspire to join, than a true community of equals. As a new user I think I'd actually be more hesitant to join a jam if everyone in there has a "senior member" or "pro jammer" badge or whatever. On a forum, publicly visible post counts and associated "levels" can give you an idea of that person's involvement and can lend weight to their contributions, which can be very thoughtful, measured and _edited_ before being posted. But on Jamulus we expose ourselves in a very different way and I couldn't care less about someone's "seniority" in the community - what I'm interested in is "can we play in sync?" and "is there a musical rapport?", which can only be determined by actually playing together. After some time, you often find you tend to meet the same people, and unless you keep changing your nick, you'll recognise each other and know what to expect. A badge is irrelevant. We already have means to state preferred genre (directories), musical ability (skill level in profile) and who we are (nick).

As for having to reintroduce yourself every time - how would a badge change that? If I'm a new user, what useful information does someone's badge convey, beyond "if you want a shiny badge like mine, you need to put in the hours" or "I've jammed here lots of times/I'm new here"? Mmmm... So what? Are you going to leave or join a server based on what someone's badge says?

Apologies if I understood this the wrong way, but that reference to commit numbers, as if that somehow invalidated @chrisrimple's opinion, is a clear example of the attitude I'd rather not provide a platform for via a badge system.

**Reply by JaminShanti** [reply]
I don't imply any outcome, other then recognize a trend.  The committal history tells me a lot of momentum occured last year.  Yes commits in an open source project implys productivity and interest.  Then I see a drop in interest this year, I base that on a reduction of commits as there is intentionally no other metics to review.  I suggest a user badge system if user retention is an issue, which is a common internet phenomenon. I didn't invent badge system, other sites use them. Post count does a similar statistic. I think you want to track the popularity of a solution or you don't. Both are fine. I have a question as I have no way to know.  After the large output of 2020 a release occurred.  What was the community like afterwards, was there a swell?  Was there no change?

**Reply by ignotus666** [reply]
I'm pretty sure the trend has much more to do with covid-related restrictions than anything else. Full lockdown in many countries led to a desperate scramble for online collaboration solutions and the user base of Jamulus exploded as a result - as did the number of contributors, for the same reason. As restrictions were lifted, people went back to physical rehearsals, but as so many people had learned about it and seen that it actually works, a good few continue to use it - far, far more than before covid. I don't think it has much to do with feeling part of a community or not, simply that people naturally prefer to play in the same physical space if they can. I don't have the numbers, but I would hazard a guess that the rise/fall in numbers of users correlates with the restrictions in place, not the timing of any release.

**Reply by pcar75** [reply]
Long Live Jamulus !

**Reply by gene96817** [reply]
I, personally, dislike the badge system. I am not a beginner and I am not an expert. Music is only part of my life so I am not an intense Jamulus user. Being an introvert, I see badges as encouraging cliques. It is natural for groups that are focused on private ensemble practice. What is interesting is how can a person know there is a welcoming space to participate.

**Reply by rdica** [reply]
Like gene96817 above, I too dislike the idea. I see no benefit, nor a problem to begin with. Working to receive badges smacks of social media behaviour and detracts from the purpose of using jamulus; to play music with others.
